compact_str/src/repr/mod.rs

```diff
@@ -13,31 +13,58 @@ mod boxed;
 mod discriminant;
 mod heap;
 mod inline;
+mod nonmax;
 mod num;
-
 mod traits;
+
 use discriminant::{
     Discriminant,
     DiscriminantMask,
 };
 use heap::HeapString;
 use inline::InlineString;
+use nonmax::NonMaxU8;
 pub use traits::IntoRepr;
 
-pub(crate) const MAX_SIZE: usize = std::mem::size_of::<String>();
-const EMPTY: Repr = Repr {
-    inline: InlineString::new_const(""),
-};
+pub const MAX_SIZE: usize = std::mem::size_of::<String>();
 
-/// Used as a discriminant to identify different variants
-pub const HEAP_MASK: u8 = 0b11111111;
+const PADDING_SIZE: usize = MAX_SIZE - std::mem::size_of::<u8>();
+const EMPTY: Repr = Repr::from_inline(InlineString::new_const(""));
 
-pub union Repr {
+/// Used as a discriminant to identify different variants
+pub const HEAP_MASK: u8 = 0b11111110;
+
+/// This is the "compiler facing" representation for the struct that underpins `CompactString`. The
+/// odd layout enables the compiler to represent an `Option<CompactString>` in the same amount of
+/// bytes as `CompactString`. In other words, it allows the compiler to see a "niche" value, which
+/// the compiler then uses to store the `None` variant.
+///
+/// The `size_of::<Repr>()` needs to be the same as `size_of::<String>()`, we construct that with
```

Comment:
```suggestion
/// The `size_of::<Repr>()` needs to be the same as `size_of::<CompactString>()`, we construct that with
```

The *needs* is to be the same size as `CompactString`. We additionally *want* the size to be the same as `String`, but this is a design choice, not a requirement.

Reply:
I referenced `String` here instead of `CompactString`, since a `CompactString` is just a wrapper around `Repr`, so the size of a `Repr` should always equal `CompactString`. Re-worded this a bit to hopefully describe our intents better